Merge branch dev/pre-commit-hook

gierle opened the pull request:
Add uniform style guide based on [conventional commits][conventional-commits] and [AngularJS][angular-commits].
Add optional tools:
- **gitlint** - enforces commit-msg style before commiting 
- **pre-commit** - runs linting, type checking and minor issue fixing

Revise pre-commit hook configuration

[angular-commits]: https://docs.google.com/document/d/1QrDFcIiPjSLDn3EL15IJygNPiHORgU1_OOAqWjiDU5Y/edit#
[conventional-commits]: https://www.conventionalcommits.org/en/v1.0.0/

gierle committed:
build: update pre-commit configuration

The defined configuration is inspired by [flake8] and [black]

[flake8]: https://github.com/PyCQA/flake8/blob/main/.pre-commit-config.yaml
[black]: https://github.com/psf/black/blob/main/.pre-commit-config.yaml

gierle committed:
build: add commit-msg config for gitlint

gierle committed:
docs: add CONTRIBUTING.md

Add style guide for commit messages
Add tool guide to enforce rules

Neonkraft commented on .pre-commit-config.yaml:
```diff
@@ -1,34 +1,54 @@
+    # According to this blog post (https://jaredkhan.com/blog/mypy-pre-commit)
+    # one should consider running mypy on all files rather than on only the committed changes
+    # since the changes can break dependencies elsewhere (e.g. function naming)
+-   repo: https://github.com/pre-commit/mirrors-mypy
```
I agree. Cleaning up the entire repo with mypy is an item on our list. We should enforce mypy in pre-commit hooks only once we've done that.

Neonkraft commented on .pre-commit-config.yaml:
```diff
@@ -1,34 +1,54 @@
+    rev: 22.12.0
+    hooks:
+    -   id: black-jupyter # supports .py and .ipynb files
+        args: [--line-length=79, --target-version=py37]
```
Optimal line length is a bitterly contested subject, but I personally find 79 to be too short. I've seen other projects that use 88. I'd much prefer that.

gierle committed:
build: update pre-commit hooks

Changes:
- line length 79->88
- stop enforcing mypy in pre-commit hooks, until the repo is cleaned using mypy
